File: challenge5/sender.cpp
#include <ros/ros.h>
#include <ros/console.h>
#include <std_msgs/Int16.h>
#include <std_msgs/Float32.h>
#include <cmath>
#include "week5_final_challenge/sender.h"

float set_point = 0.0;
double _time = 0.0;
float ref = 5;
enum CMD {
    none,
    step,
    square,
    sine
};

void define_command(int command) {
  switch (command) {
    case step:
        set_point = _time > 0 ? 1 * ref : 0;
        break;
    case square:
        set_point = std::fmod(_time*0.5,2) > 1 ? 1 * ref : -1 * ref;
        break;
    case sine:
        set_point = sin(_time*0.5)*ref;
        break;
  }
}

int main(int argc, char *argv[]) {
    ros::init(argc, argv, "sender");
    ros::NodeHandle handler;
    int nodeRate = 100;
    ros::Publisher signalPub = handler.advertise<week5_final_challenge::sender>("/set_point",10);
    ros::Rate rate(nodeRate);
    week5_final_challenge::sender set_point_out;
    int cmd = 0;
    set_point_out.set_point_data = 0.0;
    while (ros::ok()) {
        ros::param::get("/sp_type", cmd);
        _time = ros::Time::now().toSec();
        define_command(3);
        set_point_out.set_point_data = set_point;
        signalPub.publish(set_point_out);
        ros::spinOnce();
        rate.sleep();
    }
    return 0;
}

File: challenge5/sender.py
#!/usr/bin/env python

import rospy
from std_msgs.msg import Int16, Float32
from final.msg import sender
import math

set_point = 0.0
_time = 0.0
ref = 10
CMD = {
    "none": 0,
    "step": 1,
    "square": 2,
    "sine": 3
}

def define_command(command):
    global set_point, _time, ref
    if command == CMD["step"]:
        set_point = 1 * ref if _time > 0 else 0
    elif command == CMD["square"]:
        set_point = 1 * ref if math.fmod(_time*0.5,2) > 1 else -1 * ref
    elif command == CMD["sine"]:
        set_point = math.sin(_time*2) * ref

def main():
    global set_point, _time
    rospy.init_node("sender")
    nodeRate = 100
    signalPub = rospy.Publisher("/set_point", sender, queue_size=10)
    rate = rospy.Rate(nodeRate)
    set_point_out = sender()
    cmd = CMD["none"]
    set_point_out.set_point_data = 0.0
    while not rospy.is_shutdown():
        cmd = rospy.get_param("/sp_type", CMD["none"])
        _time = rospy.get_time()
        define_command(cmd)
        set_point_out.set_point_data = set_point
        signalPub.publish(set_point_out)
        rospy.spin_once()
        rate.sleep()

if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass
